### ISSUE TITLE
fastlane snapshot uninstalls on all simulators after each test instead of just one

### ISSUE THREAD

**Issue by dantman** [fastlane/fastlane#10185]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

When I run `snapshot` in a simple lane containing:
```ruby
    snapshot(
      languages: ["en-CA"],
      clear_previous_screenshots: true,
      app_identifier: "net.example.myapp",
      reinstall_app: true,
      number_of_retries: 3,
    )
```

For each device in the default list of devices being screenshot I see the roughly the following output:
```shell
[##:##:##]: Running Tests: ▸ Touching MyAppUITests.xctest
[##:##:##]: ▸ Signing {tmp}/Build/Products/Debug-iphonesimulator/MyAppUITests-Runner.app/PlugIns/MyAppUITests.xctest
[##:##:##]: ▸ All tests
[##:##:##]: ▸ Test Suite MyAppUITests.xctest started
[##:##:##]: ▸ AnonymousScreenshotTests
[##:##:##]: ▸ ✓ testLoginScreen (9.575 seconds)
[##:##:##]: ▸ AuthenticatedScreenshotTests
[##:##:##]: ▸ ✓ testInboxScreen (20.722 seconds)
[##:##:##]: ▸ 	 Executed 2 tests, with 0 failures (0 unexpected) in 30.296 (30.300) seconds
[##:##:##]: ▸ 
[##:##:##]: ▸ All tests
[##:##:##]: ▸ Test Suite MyAppTests.xctest started
[##:##:##]: ▸ 	 Executed 0 tests, with 0 failures (0 unexpected) in 0.000 (0.000) seconds
[##:##:##]: ▸ 
[##:##:##]: Collecting screenshots...
[##:##:##]: Found 2 screenshots...
[##:##:##]: Copying '{...}/fastlane/screenshots/en-CA/...-LoginScreen-....png'...
[##:##:##]: Copying '{...}/fastlane/screenshots/en-CA/...-InboxScreen-....png'...
[##:##:##]: snapshot run # of #
[##:##:##]: Patching all simulators '/Users/me/Library/Preferences/com.apple.iphonesimulator.plist' to scale to 100%
[##:##:##]: Launch Simulator iPhone 4s
[##:##:##]: $ xcrun instruments -w 51F98A72-DD48-4EF0-95C4-D32F8513BFF8 &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 51F98A72-DD48-4EF0-95C4-D32F8513BFF8 net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPhone 5
[##:##:##]: $ xcrun instruments -w AA906D32-C37F-476A-9706-3D7A9DA6C79F &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall AA906D32-C37F-476A-9706-3D7A9DA6C79F net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPad Air
[##:##:##]: $ xcrun instruments -w 79CE4238-ECD7-4C91-BA8E-270C9B159338 &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 79CE4238-ECD7-4C91-BA8E-270C9B159338 net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPhone 4s
[##:##:##]: $ xcrun instruments -w 51F98A72-DD48-4EF0-95C4-D32F8513BFF8 &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 51F98A72-DD48-4EF0-95C4-D32F8513BFF8 net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPhone 5
[##:##:##]: $ xcrun instruments -w AA906D32-C37F-476A-9706-3D7A9DA6C79F &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall AA906D32-C37F-476A-9706-3D7A9DA6C79F net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPad Air
[##:##:##]: $ xcrun instruments -w 79CE4238-ECD7-4C91-BA8E-270C9B159338 &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 79CE4238-ECD7-4C91-BA8E-270C9B159338 net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPhone 4s
[##:##:##]: $ xcrun instruments -w 51F98A72-DD48-4EF0-95C4-D32F8513BFF8 &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 51F98A72-DD48-4EF0-95C4-D32F8513BFF8 net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPhone 5
[##:##:##]: $ xcrun instruments -w AA906D32-C37F-476A-9706-3D7A9DA6C79F &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall AA906D32-C37F-476A-9706-3D7A9DA6C79F net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPad Air
[##:##:##]: $ xcrun instruments -w 79CE4238-ECD7-4C91-BA8E-270C9B159338 &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 79CE4238-ECD7-4C91-BA8E-270C9B159338 net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPad Pro
[##:##:##]: $ xcrun instruments -w C9622226-ABF2-448F-AD55-75EBDFD557FA &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall C9622226-ABF2-448F-AD55-75EBDFD557FA net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPhone 5
[##:##:##]: $ xcrun instruments -w AA906D32-C37F-476A-9706-3D7A9DA6C79F &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall AA906D32-C37F-476A-9706-3D7A9DA6C79F net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPhone 7
[##:##:##]: $ xcrun instruments -w 9C83729B-AD72-4791-BAFE-ACF5A5D4151D &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 9C83729B-AD72-4791-BAFE-ACF5A5D4151D net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPhone 7 Plus
[##:##:##]: $ xcrun instruments -w 624D086C-4676-4334-8BB6-E673B63E2DDC &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 624D086C-4676-4334-8BB6-E673B63E2DDC net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPad Air
[##:##:##]: $ xcrun instruments -w 79CE4238-ECD7-4C91-BA8E-270C9B159338 &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 79CE4238-ECD7-4C91-BA8E-270C9B159338 net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPad (5th generation)
[##:##:##]: $ xcrun instruments -w 06CE24F6-A10A-4964-AB33-BE8F7F2C2E36 &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 06CE24F6-A10A-4964-AB33-BE8F7F2C2E36 net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPad Pro (9.7 inch)
[##:##:##]: $ xcrun instruments -w 47385F91-1BBF-4F68-ABFD-F19A374BED4E &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 47385F91-1BBF-4F68-ABFD-F19A374BED4E net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPad Pro (12.9 inch)
[##:##:##]: $ xcrun instruments -w 1C0BA064-4D4A-4FF2-9588-72B685874F46 &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 1C0BA064-4D4A-4FF2-9588-72B685874F46 net.example.myapp &> /dev/null
[##:##:##]: Launch Simulator iPad Pro (10.5-inch)
[##:##:##]: $ xcrun instruments -w 43A51F04-5ACA-4442-8AC3-643A53C4C737 &> /dev/null
[##:##:##]: Uninstall application net.example.myapp
[##:##:##]: $ xcrun simctl uninstall 43A51F04-5ACA-4442-8AC3-643A53C4C737 net.example.myapp &> /dev/null
```

As you can see, after it successfully executes the tests for one device, it in turn launches every single simulator to uninstall the app (and this actually launches a new simulator and waits a bit as it partially boots up). ie: Instead of only uninstalling the app for the device that was just used or the device that'll be used next or only uninstalling on every device at the start and end of the lane, it launches every single simulator after every single test/screenshot run on a single device.

**Edit:** I stripped out the timestamps to make the output match between runs, but that process from "Patching all simulators" to when it finishes uninstalling from the last simulator takes about 4-5 minutes to complete. After each device snapshot run.

So if you are snapshotting 5 devices snapshot will: Launch A, Run test/snapshot, (Launch A, uninstall, Launch B, uninstall, Launch C, uninstall, Launch D, uninstall, Launch E, uninstall), Launch B, Run test/snapshot, (Launch A, uninstall, Launch B, uninstall, Launch C, uninstall, Launch D, uninstall, Launch E, uninstall), Launch C, Run test/snapshot, (Launch A, uninstall, Launch B, uninstall, Launch C, uninstall, Launch D, uninstall, Launch E, uninstall), Launch D, Run test/snapshot, (Launch A, uninstall, Launch B, uninstall, Launch C, uninstall, Launch D, uninstall, Launch E, uninstall), Launch E, Run test/snapshot, (Launch A, uninstall, Launch B, uninstall, Launch C, uninstall, Launch D, uninstall, Launch E, uninstall), Launch F, Run test/snapshot, (Launch A, uninstall, Launch B, uninstall, Launch C, uninstall, Launch D, uninstall, Launch E, uninstall)

Instead of sanely just doing: (Launch A, Run test/snapshot, uninstall), (Launch B, Run test/snapshot, uninstall), (Launch C, Run test/snapshot, uninstall), (Launch D, Run test/snapshot, uninstall), (Launch E, Run test/snapshot, uninstall)

### Environment

Please run `fastlane env` and copy the output below. This will help us help you :+1:
If you used `--capture_output` option please remove this block - as it is already included there.

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 10.12.6                                                        |
| Ruby                        | 2.0.0                                                          |
| Bundler?                    | true                                                           |
| Git                         | git version 2.11.0 (Apple Git-81)                              |
| Installation Source         | /usr/local/bin/fastlane                                        |
| Host                        | Mac OS X 10.12.6 (16G29)                                       |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib |
| OpenSSL Version             | OpenSSL 0.9.8zc 19 Mar 2015                                    |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version               | 8.3.3                                                          |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_CA.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
{redacted}
```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
{redacted}
```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.54.3  | ✅ Up-To-Date  |

### Loaded fastlane plugins:

| Plugin                             | Version | Update-Status |
| ---------------------------------- | ------- | ------------- |
| fastlane-plugin-versioning         | 0.2.7   | ✅ Up-To-Date  |
| fastlane-plugin-android_versioning | 0.3.0   | ✅ Up-To-Date  |

<details><summary><b>Loaded gems</b></summary>

| Gem                                | Version      |
| ---------------------------------- | ------------ |
| bundler                            | 1.14.6       |
| CFPropertyList                     | 2.3.5        |
| public_suffix                      | 2.0.5        |
| addressable                        | 2.5.2        |
| babosa                             | 1.0.2        |
| claide                             | 1.0.2        |
| colored                            | 1.2          |
| colored2                           | 3.1.2        |
| highline                           | 1.7.8        |
| commander-fastlane                 | 4.4.5        |
| declarative                        | 0.0.9        |
| declarative-option                 | 0.1.0        |
| unf_ext                            | 0.0.7.4      |
| unf                                | 0.1.4        |
| domain_name                        | 0.5.20170404 |
| dotenv                             | 2.2.1        |
| excon                              | 0.58.0       |
| multipart-post                     | 2.0.0        |
| faraday                            | 0.13.1       |
| http-cookie                        | 1.0.3        |
| faraday-cookie_jar                 | 0.0.6        |
| faraday_middleware                 | 0.12.2       |
| fastimage                          | 2.1.0        |
| gh_inspector                       | 1.0.3        |
| jwt                                | 1.5.6        |
| little-plugger                     | 1.1.4        |
| multi_json                         | 1.12.1       |
| logging                            | 2.2.2        |
| memoist                            | 0.16.0       |
| os                                 | 0.9.6        |
| signet                             | 0.7.3        |
| googleauth                         | 0.5.3        |
| httpclient                         | 2.8.3        |
| mime-types-data                    | 3.2016.0521  |
| mime-types                         | 3.1          |
| uber                               | 0.1.0        |
| representable                      | 3.0.4        |
| retriable                          | 3.1.1        |
| google-api-client                  | 0.13.4       |
| json                               | 2.1.0        |
| mini_magick                        | 4.5.1        |
| multi_xml                          | 0.6.0        |
| plist                              | 3.3.0        |
| rubyzip                            | 1.2.1        |
| security                           | 0.1.3        |
| slack-notifier                     | 1.5.1        |
| terminal-notifier                  | 1.8.0        |
| unicode-display_width              | 1.3.0        |
| terminal-table                     | 1.8.0        |
| tty-screen                         | 0.5.0        |
| word_wrap                          | 1.0.0        |
| nanaimo                            | 0.2.3        |
| xcodeproj                          | 1.5.1        |
| rouge                              | 2.0.7        |
| xcpretty                           | 0.2.8        |
| xcpretty-travis-formatter          | 0.0.4        |
| fastlane-plugin-android_versioning | 0.3.0        |
| fastlane-plugin-versioning         | 0.2.7        |
</details>

*generated on:* **2017-08-29**
</details>

**Comment by ohayon** [comment]
Hey @dantman - this is definitely a wild issue and for some reason I am not able to reproduce it on my own apps. Out of curiosity, are you running against Xcode 8 or a 9 beta?

**Comment by ohayon** [comment]
Ah! I just was able to reproduce this issue, I missed the place where you were setting `reinstall_app: true`. Seems like that param causes some issues that we should look into fixing. Just curious: is there a particular reason you are using that flag?

**Comment by dantman** [comment]
@ohayon Still using XCode 8. I want to make sure tests start with a clean install (not using the release version, not already logged in, etc...).

**Comment by dantman** [comment]
This bug is still present. The fix for this is in PR #10459, which has not been merged yet.
